src/gui/activitylistmodel.cpp

```diff
@@ -0,0 +1,231 @@
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#include <QtCore>
+#include <QAbstractListModel>
+#include <QWidget>
+#include <QIcon>
+
+#include "account.h"
+#include "accountstate.h"
+#include "accountmanager.h"
+#include "folderman.h"
+#include "accessmanager.h"
+#include "activityitemdelegate.h"
+
+#include "activitydata.h"
+#include "activitylistmodel.h"
+
+namespace OCC {
+
+ActivityListModel::ActivityListModel(QWidget *parent)
+    :QAbstractListModel(parent)
+{
+}
+
+QVariant ActivityListModel::data(const QModelIndex &index, int role) const
+{
+    Activity a;
+
+    if (!index.isValid())
+        return QVariant();
+
+    a = _finalList.at(index.row());
+    AccountStatePtr ast = AccountManager::instance()->account(a._accName);
+    QStringList list;
+
+    switch (role) {
+    case ActivityItemDelegate::PathRole:
+        list = FolderMan::instance()->findFileInLocalFolders(a._file, ast->account());
+        if( list.count() > 0 ) {
+            return QVariant(list.at(0));
+        }
+        // File does not exist anymore? Let's try to open its path
+        list = FolderMan::instance()->findFileInLocalFolders(QFileInfo(a._file).path(), ast->account());
+        if( list.count() > 0 ) {
+            return QVariant(list.at(0));
+        }
+        return QVariant();
+        break;
+    case ActivityItemDelegate::ActionIconRole:
+        return QVariant(); // FIXME once the action can be quantified, display on Icon
+        break;
+    case ActivityItemDelegate::UserIconRole:
+        return QIcon(QLatin1String(":/client/resources/account.png"));
+        break;
+    case Qt::ToolTipRole:
+    case ActivityItemDelegate::ActionTextRole:
+        return a._subject;
+        break;
+    case ActivityItemDelegate::LinkRole:
+        return a._link;
+        break;
+    case ActivityItemDelegate::AccountRole:
+        return a._accName;
+        break;
+    case ActivityItemDelegate::PointInTimeRole:
+        return Utility::timeAgoInWords(a._dateTime);
+        break;
+    case ActivityItemDelegate::AccountConnectedRole:
+        return (ast && ast->isConnected());
+        break;
+    default:
+        return QVariant();
+
+    }
+    return QVariant();
+
+}
+
+int ActivityListModel::rowCount(const QModelIndex&) const
+{
+    return _finalList.count();
+}
+
+// current strategy: Fetch 100 items per Account
+// ATTENTION: This method is const and thus it is not possible to modify
+// the _activityLists hash or so. Doesn't make it easier...
+bool ActivityListModel::canFetchMore(const QModelIndex& ) const
+{
+    if( _activityLists.count() == 0 ) return true;
+
+    QMap<AccountState*, ActivityList>::const_iterator i = _activityLists.begin();
+    while (i != _activityLists.end()) {
```

Comment:
for() loops are better since they are more readable and make it obvious it's an iteration.

 for(auto i = _activityLists.begin() ; i != _activityLists.end(); ++i) {